File: src/policies/lfu.py
"""LFU (Least Frequently Used) eviction policy.

Evicts the block with the lowest access_count. Effective for long-running
models with stable hot sets. Poor for newly-loaded blocks (cold start
problem) since new blocks always have low counts.
"""

from __future__ import annotations

from typing import TYPE_CHECKING

from .base import EvictionPolicy

if TYPE_CHECKING:
    from src.simulator.block import BlockMeta
    from src.simulator.core import GlobalState, Pool


class LFUPolicy(EvictionPolicy):
    """Evict the least frequently used block."""

    def select_victim(
        self,
        candidates: list[BlockMeta],
        pool: Pool,
        global_state: GlobalState,
    ) -> int:
        min_count = float("inf")
        victim_idx = 0
        for i, block in enumerate(candidates):
            if block.access_count < min_count:
                min_count = block.access_count
                victim_idx = i
            elif block.access_count == min_count:
                # Break ties by LRU (older block evicted first)
                if block.last_access_time < candidates[victim_idx].last_access_time:
                    victim_idx = i
        return victim_idx

    def score(
        self,
        candidates: list[BlockMeta],
        pool: Pool,
        global_state: GlobalState,
    ) -> list[float]:
        """Score by inverse frequency: less-accessed blocks get higher scores."""
        if not candidates:
            return []
        max_count = max(b.access_count for b in candidates)
        min_count = min(b.access_count for b in candidates)
        span = max_count - min_count if max_count != min_count else 1
        return [
            (max_count - b.access_count) / span
            for b in candidates
        ]

    def name(self) -> str:
        return "LFU"

File: slm_os_integration/src/lfu.rs
//! Least-Frequently-Used eviction.
//!
//! Mirrors `src/policies/lfu.py`: pick the candidate with the smallest
//! `access_count`; ties are broken by older `last_access_time` (LRU
//! within the tied set).

use alloc::vec::Vec;

use crate::block::BlockMeta;
use crate::eviction_policy::EvictionPolicy;

extern crate alloc;

#[derive(Default)]
pub struct LfuPolicy;

impl LfuPolicy {
    pub fn new() -> Self {
        Self
    }
}

impl EvictionPolicy for LfuPolicy {
    fn select_victim(&mut self, candidates: &[BlockMeta]) -> usize {
        debug_assert!(!candidates.is_empty(), "LFU select_victim on empty list");
        let mut victim = 0;
        let mut min_count = candidates[0].access_count;
        for (i, c) in candidates.iter().enumerate().skip(1) {
            if c.access_count < min_count {
                min_count = c.access_count;
                victim = i;
            } else if c.access_count == min_count
                && c.last_access_time < candidates[victim].last_access_time
            {
                victim = i;
            }
        }
        victim
    }

    fn score(&mut self, candidates: &[BlockMeta]) -> Vec<f32> {
        if candidates.is_empty() {
            return Vec::new();
        }
        let mut min_count = candidates[0].access_count;
        let mut max_count = candidates[0].access_count;
        for c in &candidates[1..] {
            if c.access_count < min_count { min_count = c.access_count; }
            if c.access_count > max_count { max_count = c.access_count; }
        }
        let span = if max_count == min_count { 1 } else { max_count - min_count };
        candidates
            .iter()
            .map(|c| (max_count - c.access_count) as f32 / span as f32)
            .collect()
    }

    fn name(&self) -> &'static str {
        "LFU"
    }
}
